### ISSUE TITLE
MySQL does not support millisecond accuracy in timestamps (problem for JobParameters) [BATCH-1379]

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-batch#2202]
**[Dave Syer](https://jira.spring.io/secure/ViewProfile.jspa?name=david_syer)** opened **[BATCH-1379](https://jira.spring.io/browse/BATCH-1379?redirect=false)** and commented

MySQL does not support millisecond accuracy in timestamps (problem for JobParameters).  The issue is that round tripping from JobParameters to the database and back does not preserve the timestamp, so a job can be launched twice with different parameters, when they should have been the same.

---

**Affects:** 2.1.0

**Reference URL:** http://forum.springsource.org/showthread.php?t=74219

**Issue Links:**
- [BATCH-2522](https://jira.spring.io/browse/BATCH-2522) Job key hash returns different result for restart when using date parameter
 (_**"is duplicated by"**_)
- [BATCH-2542](https://jira.spring.io/browse/BATCH-2542) Multiple Job Intance for same job parameters
 (_**"is duplicated by"**_)

**Comment by benas** [comment]
I was able to reproduce the issue with Spring Batch 4.2.1 and MySQL 8.0.18 (See [this repo](https://github.com/benas/spring-batch-sandbox/tree/master/gh2202)).

MySQL added support for milliseconds accuracy since v5.6.4 (released in 2011): https://dev.mysql.com/doc/relnotes/mysql/5.6/en/news-5-6-4.html#mysqld-5-6-4-fractional-seconds

The issue is that the DDL that Spring Batch provides for MySQL does not declare the fractional seconds precision for the [BATCH_JOB_EXECUTION_PARAMS#DATE_VAL](https://github.com/spring-projects/spring-batch/blob/master/spring-batch-core/src/main/resources/org/springframework/batch/core/schema-mysql.sql#L32) column.

Using the fsp syntax as described in MySQL docs fixes the issue:

```diff
CREATE TABLE BATCH_JOB_EXECUTION_PARAMS  (
     JOB_EXECUTION_ID BIGINT NOT NULL ,
     TYPE_CD VARCHAR(6) NOT NULL ,
     KEY_NAME VARCHAR(100) NOT NULL ,
     STRING_VAL VARCHAR(250) ,
--   DATE_VAL DATETIME DEFAULT NULL ,
++   DATE_VAL DATETIME(6) DEFAULT NULL ,
     LONG_VAL BIGINT ,
     DOUBLE_VAL DOUBLE PRECISION ,
     IDENTIFYING CHAR(1) NOT NULL ,
     constraint JOB_EXEC_PARAMS_FK foreign key (JOB_EXECUTION_ID)
     references BATCH_JOB_EXECUTION(JOB_EXECUTION_ID)
) ENGINE=InnoDB;
```

**Comment by parikshitdutta** [comment]
Hi @benas, is it open for contribution? I can take it.

**Comment by benas** [comment]
@parikshitdutta Everything without a PR is open for contribution, so you are welcome! For this one we need to:

1. Update the DDL for MySQL to add the fractional seconds precision as mentioned in my [previous comment](https://github.com/spring-projects/spring-batch/issues/2202#issuecomment-574237266) (for new users)
2. Add a [migration script](https://github.com/spring-projects/spring-batch/tree/master/spring-batch-core/src/main/resources/org/springframework/batch/core/migration) for v4.3 with the statement that modifies the column `ALTER TABLE MODIFY COLUMN ...` (for existing users)
3. Add a test to cover the case

Point 3) is a bit tricky to do automatically, so I will do it manually for now. If you want to work on this issue, you can open a PR with 1) and 2).

Now here is the reason why testing this automatically is a bit tricky. Until now, we have been using embedded DBs and messages Brokers in most of our automated test suite. However, some parts are tested manually because it is impossible to embed the target database or broker in a JVM (MySQL is one of them). I am preparing a test infrastucture based on Docker + testcontainers. Once this test infrastructure is ready, we can add an automated test that covers this bug. But for now, I will test the fix manually.

**Comment by parikshitdutta** [comment]
Hi @benas, as discussed except the test case to cover the scenario, other changes done and PR#3704 opened for your review.

Thank you.

**Comment by benas** [comment]
Fixed with #3704 . @parikshitdutta Thank you for your contribution!
